### ISSUE TITLE
Ability to define arguments as well (like flags)

### ISSUE THREAD

**Issue by andygrunwald** [spf13/cobra#378]
Thanks for cobra. It is a really incredible project. Kudos.

In the PHP language there is a similar library quite popular: [symfony/console](https://github.com/symfony/console).
In this library you are able to define flags and _arguments_ and access them via names.
This makes development quite readable and convenient.

In cobra arguments are treated as a slice / list instead of defined names. See [cobra/cmd/add.go#L47](https://github.com/spf13/cobra/blob/master/cobra/cmd/add.go#L47) as an example.

In `symfony/console` you would write something like

```php
protected function configure()
{
    $this
       // ...
        ->addArgument('name', InputArgument::REQUIRED, 'Who do you want to greet?')
        ->addArgument('last_name', InputArgument::OPTIONAL, 'Your last name?')
    ;
}
```

and access it like

```php
protected function execute(InputInterface $input, OutputInterface $output)
{
    $text = 'Hi '.$input->getArgument('name');

    $lastName = $input->getArgument('last_name');
    if ($lastName) {
        $text .= ' '.$lastName;
    }

    $output->writeln($text.'!');
}
```

A matching call would look like `php appname Andy Grunwald`.

Did you considered such a behaviour for cobra as well?
What is your opinion about it?

I don`t know much about cobras internal codebase (yet), but i assume this wouldnt be a BC.
I like to hear your opinion and feedback about it.

A few documentation references that might be useful for this:
* [Symfony/Console docs](https://symfony.com/doc/current/components/console.html)
* [Console Input (Arguments & Options) @ Symfony/Console docs](https://symfony.com/doc/current/console/input.html)
* [Understanding how Console Arguments Are Handled @ Symfony/Console docs](https://symfony.com/doc/current/components/console/console_arguments.html)

**Comment by bogem** [comment]
Some progress was made in #284

**Comment by Naatan** [comment]
While relevant, I don't think #284 is the same as this. This asks for a way to properly specify and document arguments, and likely some type of automation of what #284 added.

**Comment by ceejatec** [comment]
Agreed. I'd go one step further and say there should be the ability to validate types of positional arguments as well. I can already specify that my command takes a --foo argument whose value is of type int, and cobra/pflags will validate the user's input, convert it to an int, and bind it to a corresponding variable. But if I want it as a positional argument I have to roll my own everything.

**Comment by ceejatec** [comment]
Python's built-in argparse library handles this very smoothly. There's no difference at all between a positional and flag-based argument except that one is named eg. "foo" and the other is named "--foo". Everything else - typing, validation, number of values, etc - is exactly the same. And the auto-generated help output documents them all correctly.

**Comment by hehnope** [comment]
Didn't want to create a new issue; but to be clear, there is no support to set "default" values for a flag? Not complaining, just don't want to open another issue.

**Comment by alecb-stripe** [comment]
++ Following argparse's lead on positional arguments.

**Comment by mlaforet** [comment]
Just started using this package today
I was having so much fun until I ran into this :(

**Comment by johnSchnake** [comment]
Consistent activity even over a few years with some thoughtful discussion. I think this warrants some extra thought but would definitely need a full proposal as it seems like a potentially large change and the implementation that can conserve backwards compatibility isn't clear.

**Comment by gwynforthewyn** [comment]
I was just trying to figure out how to bind arguments to variable names and came to this issue, so I wanted to add a +1 to show interest.

**Comment by rmonvfer** [comment]
Hello! I know this has been open for some time but I'd like to bump this a little. I find it truly baffling that there is no way to document positional arguments, it feels like something pretty foundational to CLI apps...
